Fixed issue of notes being inside repeat block in phrase maker

meganindya opened the pull request:
@pikurasa This fixes #2051. The repeats were being skipped because the IDs of note blocks inside repeat blocks were getting augmented and so weren't producing a match.<br/>
For example, in
![x2](https://user-images.githubusercontent.com/33328728/77794998-d3d96f80-7092-11ea-9ff1-d90ec7843338.png)
the block IDs were like
![x1](https://user-images.githubusercontent.com/33328728/77795028-e0f65e80-7092-11ea-8071-5ceea4aa65f2.png)
The IDs were getting added to a larger number. I fixed the comparison by comparing the trailing characters only, which would then match.<br/>
I've removed the sort when the phrase maker matrix is created and it contains note blocks. This is because we might want to see the corresponding note in the matrix when the note is part of some other block, e.g. `action`. Also, this could be a step towards being able to change the corresponding note from the phrase maker itself, which would be reflected on the blocks. However, this feature isn't added yet. Irrespective, we can sort the matrix by clicking on the sort button anytime.<br/>
In addition, I've fixed an invalid access case which was producing error.

walterbender commented:
The large number was added as a work around to repeats being used when not in a note block: it was necessary to give each iteration in the loop a unique id in order to get the "restored" state to process properly when reopening the widget. Not sure what the correct behavior should be for note blocks. In any case, I will review this PR. Thx

walterbender commented:
It does a nice job with the repeat. But it doesn't handle restoring changes to matrix properly -- all of the updates get associated with the first iteration of the loop.

walterbender commented:
![Screenshot from 2020-03-27 20-06-58](https://user-images.githubusercontent.com/3996641/77809601-bd1d2380-7066-11ea-97ca-411fe8d520a4.png)

![Screenshot from 2020-03-27 20-06-32](https://user-images.githubusercontent.com/3996641/77809604-c1e1d780-7066-11ea-8e16-dafee8d130a1.png)

![Screenshot from 2020-03-27 20-08-19](https://user-images.githubusercontent.com/3996641/77809608-c7d7b880-7066-11ea-9dd0-cb86ff81ae66.png)

meganindya commented:
Yes, I can successfully reproduce this. Also, the new notes don't get reflected in the blocks. However, as I've mentioned, I removed the sort so every note is reflected in the matrix. Previously, there was the sort so duplicates would be removed (although it wasn't working due to the invalid access that I fixed).

The way I see this, we shouldn't allow users to make any changes in the matrix if the `_noteBlocks` variable is found to be true, i.e. there exists one or more note blocks while creating the matrix. I say this because it can lead to several anomalies.

For example, in the example I've provided above, if I'd initially have F4, F4, G4, F4, F4, G4, F4, F4, G4, F4, F4, G4. Here, it's such that F4, F4, G4 is repeated 4 times (part of 'Sakura' action block). So, if I were to change say the first F4 to C4, I'd require a change in each. Again, for example, in the case you've provided, if I add new notes, I wouldn't know exactly in which block they are to be added. I think it is essential that the blocks and the matrix reflect the same thing. In complex cases, I think it can be totally anomalous, hence I believe we should block changes in the matrix if `_noteBlocks` is true.

meganindya commented:
Yes, I can successfully reproduce this. As I mentioned, there's no feature for the new notes to get reflected in the blocks. However, as I've mentioned, I removed the sort so every note is reflected in the matrix. Previously, there was the sort so duplicates would be removed (although it wasn't working due to the invalid access that I fixed).

The way I see this, we shouldn't allow users to make any changes in the matrix if the `_noteBlocks` variable is found to be true, i.e. there exists one or more note blocks while creating the matrix. I say this because it can lead to several anomalies.

For example, in the example I've provided above, if I'd initially have F4, F4, G4, F4, F4, G4, A4, A4, B4, A4, A4, B4. Here, it's such that F4, F4, G4 is repeated twice (part of 'Sakura' action block scalar-transposed), and A4, A4, B4 is repeated twice (part of 'Sakura' action block unchanged). So, if I were to change say the first F4 to C4, I'd require a change inside the action block itself. It can get messy like in this case, since the action block is inside a scalar-transpose block once, and outside of it in another.

Again, for example, in the case you've provided, if I add new notes, I wouldn't know exactly in which block they are to be added. I think it is essential that the blocks and the matrix reflect the same thing.

Also, if I add a new pitch in the matrix, I wouldn't be sure if I should add a note block or a pitch and rhythm blocks once its columns start populating. Again, changes in the note-values would be impossible to reflect in such cases.

In complex cases, I think it can be totally anomalous, hence I believe we should block changes in the matrix if `_noteBlocks` is true.

walterbender commented:
The way I got around this in the non-note situation was to keep a list of grid settings per block (the list was maintained by adding the "large number".

But I think we should allow changes, but those changes would only be seen when you export. They would not be preserved between sessions.

meganindya commented:
But, by doing so, the matrix and the blocks wouldn't reflect the same thing. Would that be right?

walterbender commented:
Not sure I follow. The export from the matrix would match the grid. The only problems I see are:
(1) with note blocks, if you reopen the matrix, you won't see the changes you made previously.
(2) if you add/change notes in the grid, you won't see them in the blocks used to create the grid.

#2 should be true whenever a repeat block is used.

meganindya commented:
To be a little verbose, the expected behaviour is:
(1) with note blocks, if I make some changes to the matrix, close it and reopen, those changes will be gone.
(2) with note blocks, if I add/change notes in the matrix, those changes will not be reflected in the blocks.<br/>
However, exported grid will match the matrix. The export feature is working correctly. Basically, every time the matrix is opened, the original representation from the block stack will be restored; new changes will be lost.<br/>
Am I right?

walterbender commented:
Yes. I think that is a reasonable behavior.

meganindya commented:
I suppose this could be a new issue then. I'll try to fix it.

walterbender commented:
Yes. Makes sense as a new issue.
